core/trino-main/src/main/java/io/trino/execution/DropColumnTask.java

```diff
@@ -87,17 +91,53 @@ public ListenableFuture<Void> execute(
             return immediateVoidFuture();
         }
 
-        if (metadata.getColumnMetadata(session, tableHandle, columnHandle).isHidden()) {
+        List<String> field = statement.getColumn().getOriginalParts().stream().map(Identifier::getValue).skip(1).collect(toImmutableList());
+        ColumnMetadata columnMetadata = metadata.getColumnMetadata(session, tableHandle, columnHandle);
+        if (columnMetadata.isHidden()) {
             throw semanticException(NOT_SUPPORTED, statement, "Cannot drop hidden column");
         }
-
-        if (metadata.getTableMetadata(session, tableHandle).getColumns().stream()
-                .filter(info -> !info.isHidden()).count() <= 1) {
-            throw semanticException(NOT_SUPPORTED, statement, "Cannot drop the only column in a table");
+        if (field.isEmpty()) {
+            if (metadata.getTableMetadata(session, tableHandle).getColumns().stream()
+                    .filter(info -> !info.isHidden()).count() <= 1) {
+                throw semanticException(NOT_SUPPORTED, statement, "Cannot drop the only column in a table");
+            }
+            metadata.dropColumn(session, tableHandle, columnHandle);
+        }
+        else {
+            if (columnMetadata.getType() instanceof RowType rowType) {
+                if (findField(statement, rowType.getFields(), field).isEmpty()) {
+                    if (!statement.isColumnExists()) {
+                        throw semanticException(COLUMN_NOT_FOUND, statement, "Field '%s' does not exist", statement.getColumn());
+                    }
+                    return immediateVoidFuture();
+                }
+            }
+            else {
+                // TODO: Support array and map types
+                throw semanticException(NOT_SUPPORTED, statement, "Dropping fields from %s is not supported", columnMetadata.getType());
+            }
+            metadata.dropField(session, tableHandle, columnHandle, field);
         }
-
-        metadata.dropColumn(session, tableHandle, columnHandle);
 
         return immediateVoidFuture();
     }
+
+    private Optional<RowType.Field> findField(DropColumn statement, List<RowType.Field> fields, List<String> fieldPath)
+    {
+        String fieldName = fieldPath.get(0);
+        List<String> childField = fieldPath.subList(1, fieldPath.size());
+
+        if (fields.stream().allMatch(field -> field.getName().orElseThrow().equals(fieldName))) {
+            throw semanticException(NOT_SUPPORTED, statement, "Cannot drop the only field in a row type");
```

Comment:
counting the fields would be more direct way of checking whether there is exactly one
(dropField should not drop multiple fields, even if they somehow have the same name)

also, there is nothing preventing the connector to have unnamed fields, so `orElseThrow` isn't good (Cassandra tuples?)

also, application of this logic isn't expected in a "find field" method. 
also, top-level ROW can have single field, that's OK. only the leaf field (the one removed) needs to have a sibling

Reply:
> dropField should not drop multiple fields, even if they somehow have the same name

Could you explain the rationale? If the connector supports adding such fields, allowing dropping the duplicated fields makes sense to me.

Reply:
I, as a user, expect single "DROP field" statement to (either fail or) drop one field.
I do not expect the statement to drop multiple fields.